ext/standard/array.c

```diff
@@ -388,6 +389,33 @@ static int php_array_reverse_data_compare(const void *a, const void *b) /* {{{ *
 }
 /* }}} */
 
+static int php_array_data_identical(const void *a, const void *b)
+{
+	Bucket *f;
+	Bucket *s;
+	zval *first;
+	zval *second;
+
+	f = (Bucket *) a;
+	s = (Bucket *) b;
+
+	first = &f->val;
+	second = &s->val;
+
+	if (UNEXPECTED(Z_TYPE_P(first) == IS_INDIRECT)) {
+		first = Z_INDIRECT_P(first);
+	}
+	if (UNEXPECTED(Z_TYPE_P(second) == IS_INDIRECT)) {
+		second = Z_INDIRECT_P(second);
+	}
+	return zend_is_identical(first, second) == 1 ? 0 : 1;
```

Comment:
Maybe
```suggestion
	return zend_is_identical(first, second) != 1;
```
?

Reply:
I'm in favor of having a SORT_IDENTICAL flag that would impose a strict ordering that was transitive. SORT_STRICT or SORT_MACHINE would be other possible names (they seem more descriptive if they get passed to `sort()` - SORT_IDENTICAL doesn't describe what order they'd end up in)

1. **This doesn't sort values properly.** It'd work if the values were adjacent, but it wouldn't work if the values weren't adjacent (e.g. `[$o1, $o2, $o3, $o1, $o2, $o3]` will almost certainly fail for deduplication, where the values are different stdClass instances. It'd have to return `-1` some of the time (based on some criteria) to ensure that identical values became adjacent.

   It probably also fails for `array_unique([1, 2, 3, 0, 1, 2, 3], SORT_IDENTICAL)`, but I haven't checked - that would be a useful test case to add.
2. I'd suggest falling back to the SORT_REGULAR criteria, but I think it has some of it's own issues and you might want to write a brand new sorting helper
3. ZVAL_DEREF to convert IS_REFERENCE to a regular value after converting IS_INDIRECT
4. One way to break ties is to first sort by Z_TYPE_P(first) and Z_TYPE_P(second), then perform a type-based comparison to break ties

    - You might want to treat IS_LONG and IS_DOUBLE as the same position so that `sort($array, SORT_IDENTICAL)` will work intuitively and put `[0, 0.1, 0.5, 1]` in the expected order, if this flag ends up used for sorting.
5. After sorting by Z_TYPE_P, everything would be relatively easy except arrays. You'd need to distinguish `["0"]` from `[0]` in `[[0], ["0"], [0], "0"]`.

    Objects could be compared by object id and optionally class name; resources by resource id, etc.
6. The float `NAN` is an annoying edge case, but probably out of scope to handle, since `NAN <= NAN` is false.